module/zfs/zio_compress.c

```diff
@@ -143,8 +143,18 @@ zio_compress_data(enum zio_compress c, abd_t *src, void *dst, size_t s_len,
 	if (c == ZIO_COMPRESS_EMPTY)
 		return (s_len);
 
-	/* Compress at least 12.5% */
-	d_len = s_len - (s_len >> 3);
+	/*
+	 * Usable compression thresholds are:
+	 * - less than BPE_PAYLOAD_SIZE (embedded_data feature)
+	 * - at least one saved sector
+	 */
+	if (compress_threshold > 0) {
+		d_len = MAX(BPE_PAYLOAD_SIZE,
+		    s_len - (s_len % compress_threshold));
```

Comment:
It looks like this would produce an undesirable d_len (albeit in the safe(?) 0..s_len range) if s_len > compress_threshold.  And I can't quickly see/intuit a reason why s_len > compress_threshold is always true; if it is, then please add a comment. :)

Reply:
Wondering if this is more readable / provable:
```c
	/*
	 * Usable compression thresholds are:
	 * - less than BPE_PAYLOAD_SIZE (embedded_data feature)
	 * - at least one saved sector
	 */
	if (compress_threshold > 0) {
		if (s_len > compress_threshold)
			d_len = MAX(BPE_PAYLOAD_SIZE, s_len - compress_threshold);
		else
			d_len = BPE_PAYLOAD_SIZE;

	} else {
		/* Special case for reproducibility of ARC checks */
		d_len = s_len;
	}
```

Reply:
What do you mean by `undesirable`? Could you please give an example?

`s_len - (s_len % compress_threshold)` will give d_len rounded to sector size (because compress_threshold is a max sector size if this incarnation of PR).

And I don't like additional `if`, looks redundant for me.